### PR TITLE
feat: CollectingAuditSink with local_sink property

### LINKED ISSUE #70: from_server() should preserve local audit trail via TeeAuditSink
## Problem

`Edictum.from_server()` replaces the audit sink entirely with a `ServerAuditSink` that sends events to the console. The local process loses all visibility into governance decisions — there's no way to inspect what just happened without querying the server.

This breaks:
- **Adapter demos**: `classify_result()` relies on `sink.since_mark()` to determine if a call was denied, redacted, or observed. In console mode, every scenario shows as ALLOWED or AGENT_SKIP because there are no local events.
- **Programmatic inspection**: Any code that needs to react to governance decisions post-hoc (e.g., logging, conditional logic, testing) can't access audit events.
- **Callbacks**: `on_postcondition_warn` works, but there's no equivalent for pre-decision inspection.

## Expected Behavior

`from_server()` should keep a local audit trail AND send to the server. The server sink should be additive, not a replacement.

## Proposed Solution

Introduce a `TeeAuditSink` that writes to both a local `CollectingAuditSink` and the `ServerAuditSink`:

```python
class TeeAuditSink:
    def __init__(self, *sinks):
        self.sinks = sinks
        self.local = next((s for s in sinks if isinstance(s, CollectingAuditSink)), None)

    async def emit(self, event):
        for sink in self.sinks:
            await sink.emit(event)
```

Then `from_server()` uses it internally:

```python
local_sink = CollectingAuditSink()
server_sink = ServerAuditSink(url, api_key)
guard.audit_sink = TeeAuditSink(local_sink, server_sink)
```

Callers can then always access `guard.audit_sink.local.since_mark()` regardless of mode.

## Relationship to Other Issues

- Depends on / part of the `public guard.audit_sink` work (v0.14 plan)
- Would fix console-mode classification for all 8 adapter demos
- Related to #67 (cross-event-loop) since the server sink's async httpx client is what breaks sync adapters

## Impact

All adapter demos in console mode (`--console`) currently lose governance result classification. Verified with: LangChain (partial), OpenAI Agents, Semantic Kernel, Google ADK, Claude Agent SDK. The governance runs correctly server-side, but the demo output is misleading.

### PR DESCRIPTION
**Pull request by acartag7**
## Summary

- Adds `CollectingAuditSink` — in-memory audit sink with bounded ring buffer (50k default) and mark-based windowed queries
- Adds `Edictum.local_sink` property — always-present `CollectingAuditSink` on every instance regardless of construction method
- Adds `MarkEvictedError` exception for evicted marks, `ValueError` for future marks
- **Breaking**: `Edictum()` no longer adds `StdoutAuditSink` by default. Stdout output requires explicit `audit_sink=StdoutAuditSink()` or YAML `observability.stdout: true`
- When `audit_sink` is provided (single or list), wrapped in `CompositeSink` with `local_sink` first

Closes #70 — console mode now has local audit trail via `local_sink`

## Test plan

- [x] 18 unit tests for `CollectingAuditSink` (mark/since_mark, eviction, overflow, clear, filter, security)
- [x] 12 integration tests for `local_sink` (all construction paths + server mode)
- [x] All 1607 existing tests pass
- [x] Lint clean, docs build clean
- [x] edictum-demo standalone: 7/7 PASS (5D 1R 1O 5A across all adapters)
- [x] edictum-demo console mode: 4 PASS, 3 CHECK (LLM variance, governance working)

<h3>Greptile Summary</h3>

This PR adds `CollectingAuditSink` — an in-memory, bounded ring-buffer audit sink with mark-based windowed queries — and exposes it as the always-present `Edictum.local_sink` property on every instance regardless of construction path. It also carries a documented breaking change: `Edictum()` no longer silently outputs JSON to stdout by default; stdout logging now requires an explicit opt-in.

Key changes:
- **`src/edictum/audit.py`**: New `CollectingAuditSink` class and `MarkEvictedError` exception. The mark/eviction arithmetic is correct, but the ring-buffer truncation uses `self._events[-self._max_events:]`, which silently breaks when `max_events=0` because `-0 == 0` in Python — the buffer grows unboundedly. Input validation (`max_events >= 1`) is missing.
- **`src/edictum/__init__.py`**: `_local_sink` is initialised before the `audit_sink` branching, and is always inserted first in any `CompositeSink`. The `local_sink` property is clean and correctly typed. Breaking-change migration note in CHANGELOG is clear.
- **Tests**: Good coverage across all construction paths and edge cases. Two minor issues in `test_collecting_sink_behavior.py`: an `id()`-based defensive-copy assertion that is unreliable under CPython address reuse, and a dead `sink.mark()` call whose return value is discarded.

<h3>Confidence Score: 3/5</h3>

- Safe to merge after fixing the `max_events=0` ring-buffer silent-failure bug; test issues are minor but the `id()` assertion can give false confidence.
- The core logic for marks, eviction boundaries, and composite-sink composition is correct. The one real bug — passing `max_events=0` (or a negative value) causes the ring buffer to grow without bound, silently defeating the OOM-prevention guarantee — lowers confidence. Two test-quality issues (unreliable `id()` check, dead `mark()` call) don't affect production behaviour but reduce the value of the test suite. Everything else, including the breaking-change documentation and migration note, is well-executed.
- `src/edictum/audit.py` — missing `max_events >= 1` guard in `CollectingAuditSink.__init__`; `tests/test_behavior/test_collecting_sink_behavior.py` — unreliable identity assertion and dead mark call.

<details><summary><h3>Important Files Changed</h3></summary>

| Filename | Overview |
|----------|----------|
| src/edictum/audit.py | Adds `CollectingAuditSink` and `MarkEvictedError`. The mark/eviction logic is correct, but `max_events=0` (or negative) silently breaks the ring buffer due to Python's `-0 == 0` semantics — the buffer grows unboundedly instead of being capped. |
| src/edictum/__init__.py | Adds `_local_sink` / `local_sink` property to `Edictum.__init__`, correctly composing it first in `CompositeSink` for all construction paths. Breaking change (no more default `StdoutAuditSink`) is clearly documented. |
| tests/test_behavior/test_collecting_sink_behavior.py | Good coverage of marks, eviction, overflow, clear, and filter. Two issues: the `id()`-based defensive-copy assertion is unreliable (CPython address reuse), and there's a dead `sink.mark()` call in `test_clear_invalidates_pre_clear_marks`. |
| tests/test_behavior/test_local_sink_behavior.py | Thorough integration tests for `local_sink` across all construction paths. `tmp_path` fixture correctly used for YAML temp files (previous leak fixed). No issues found. |
| tests/test_server/test_from_server_local_sink.py | New server-mode integration tests verify `local_sink` presence, event receipt, custom sink tee-ing, and composite sink structure. Mocks are correct and tests are well-scoped. |

</details>

</details>

<h3>Sequence Diagram</h3>

```mermaid
sequenceDiagram
    participant Caller
    participant Edictum
    participant CompositeSink
    participant CollectingAuditSink as local_sink (CollectingAuditSink)
    participant UserSink as user_sink (optional)

    Caller->>Edictum: Edictum(audit_sink=user_sink)
    Edictum->>CollectingAuditSink: _local_sink = CollectingAuditSink()
    Edictum->>CompositeSink: audit_sink = CompositeSink([local_sink, user_sink])

    Caller->>CollectingAuditSink: m = guard.local_sink.mark()
    CollectingAuditSink-->>Caller: m = _total_emitted

    Caller->>Edictum: await guard.run("tool", args, fn)
    Edictum->>CompositeSink: await audit_sink.emit(pre_event)
    CompositeSink->>CollectingAuditSink: await emit(pre_event)
    CollectingAuditSink->>CollectingAuditSink: append + evict if overflow
    CompositeSink->>UserSink: await emit(pre_event)

    Edictum->>CompositeSink: await audit_sink.emit(post_event)
    CompositeSink->>CollectingAuditSink: await emit(post_event)
    CompositeSink->>UserSink: await emit(post_event)

    Caller->>CollectingAuditSink: events = guard.local_sink.since_mark(m)
    alt mark still valid
        CollectingAuditSink-->>Caller: [pre_event, post_event]
    else mark evicted (buffer overflow)
        CollectingAuditSink-->>Caller: raises MarkEvictedError
    end
```

<details><summary><h3>Comments Outside Diff (2)</h3></summary>

1. `tests/test_behavior/test_collecting_sink_behavior.py`, line 517-520 ([link](https://github.com/acartag7/edictum/blob/6933f620dac49917f14d6a476673f417a198ade9/tests/test_behavior/test_collecting_sink_behavior.py#L517-L520)) 

   **Unreliable `id()` identity check**

   The assertion `assert id(sink.events) != id(sink.events)` is not a reliable way to verify a defensive copy. In CPython, the first temporary list is immediately garbage-collected before the second call, making it eligible for memory reuse — so the two calls can return the same address and the assert can silently pass even if the property returned the same object both times. This is technically flaky.

   Assigning both results to named variables prevents premature GC and makes the intent explicit:

2. `tests/test_behavior/test_collecting_sink_behavior.py`, line 467-475 ([link](https://github.com/acartag7/edictum/blob/6933f620dac49917f14d6a476673f417a198ade9/tests/test_behavior/test_collecting_sink_behavior.py#L467-L475)) 

   **Unused `mark()` return value — dead code**

   `sink.mark()` is called on line 471 but its return value is discarded. The test then uses the hard-coded literal `3` in `since_mark(3)`, making the `mark()` call do nothing. This is confusing because it implies the mark taken here is the one being validated, when in fact the magic number `3` is what drives the assertion.

   Consider either storing and using the captured mark, or removing the dead call:

</details>

<sub>Last reviewed commit: 6933f62</sub>

> Greptile also left **1 inline comment** on this PR.

### CODE REVIEW

File: tests/test_behavior/test_local_sink_behavior.py [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+      effect: deny
+      message: "never fires"
+"""
+
+
+def _write_yaml(content: str = MINIMAL_YAML) -> Path:
+    f = tempfile.NamedTemporaryFile(suffix=".yaml", mode="w", delete=False)
+    f.write(content)
+    f.close()
+    return Path(f.name)
```
[the remark this answers is not in the document]
Comment (acartag7) [reply tests/test_behavior/test_local_sink_behavior.py]: Fixed in 6933f62 — replaced `_write_yaml()` with pytest `tmp_path` fixture for automatic cleanup.
